packages/lesswrong/lib/helpers.ts

```diff
@@ -11,6 +11,21 @@ export const messageGetLink = (message: DbMessage): string => {
   return `/inbox/${message.conversationId}`;
 };
 
+
+export function constantTimeCompare(a: string, b: string) {
+  try {
+    const firstArr = a.split('');
+    const secondArr = b.split('');
+
+    const allCharsEqual = firstArr.every((firstArrChar, idx) => secondArr[idx] === firstArrChar);
```

Comment:
I don't think this is constant time? I'm not sure where `Array.every` comes from, btu the normal way to implement it would involve an early-exit on the first mismatched character.

Reply:
The rules for constant time are:
* No `if` or `?:` based on the data
* No calling functions whose identity varies based on the data
* No calling functions which themselves break either of these rules
I checked, and `Array.every` does indeed have early exit.

Reply:
I think the right answer is probably not to try to write a correct constant-time comparison function, it's to do the wiring so that code in `lib` can call it (if it's on the server) or call something else (if it's on the client).

Reply:
(For context, I was working on this with @Discordius and pushed the PR from his account for technical reasons.)

Huh, I'm not sure why I thought `Array.every` didn't have early exit, since I've explicitly relied on its short-circuiting behavior in the past 🤦.  I think we technically don't need constant time, we just need time to not change with respect to anything except the input length, so a regular `for` loop iterating over the full length of the input should work fine.

w.r.t. having separate logic for client- and server-side comparisons - if there's a standard way to do this in native JS, then it seems preferable to use the same mechanism for comparisons regardless of location.  If there isn't, it sounds like we need to write one for the client anyways.  It seems better to have one function responsible for comparing the `linkSharingKey` rather than two, all else equal.  Is the concern that a home-rolled function would be less secure than using the native `crypto` implementation, so even if we can't use it in the parts of code that at "client-side", we should still use it on the server, to avoid exposing any unnecessary surface area, even at the cost of having two functions doing more or less the same thing?  Can discuss further tomorrow.